mobile/lib/widgets/asset_grid/thumbnail_image.dart

```diff
@@ -223,20 +86,247 @@ class ThumbnailImage extends ConsumerWidget {
+            duration.format(),
+            style: const TextStyle(
+              color: Colors.white,
+              fontSize: 10,
+              fontWeight: FontWeight.bold,
             ),
           ),
-      ],
+          const SizedBox(width: 3),
+          const Icon(
+            Icons.play_circle_fill_rounded,
+            color: Colors.white,
+            size: 18,
+          ),
+        ],
+      ),
+    );
+  }
+}
+
+class StackIcon extends StatelessWidget {
+  final bool isVideo;
+  final int stackCount;
+
+  const StackIcon({super.key, required this.isVideo, required this.stackCount});
+
+  @override
+  Widget build(BuildContext context) {
+    return Positioned(
+      top: isVideo ? 28 : 5,
+      right: 8,
+      child: Row(
+        children: [
+          if (stackCount > 1)
+            Text(
+              "$stackCount",
+              style: const TextStyle(
+                color: Colors.white,
+                fontSize: 10,
+                fontWeight: FontWeight.bold,
+              ),
+            ),
+          if (stackCount > 1)
+            const SizedBox(
+              width: 3,
+            ),
+          const Icon(
+            Icons.burst_mode_rounded,
+            color: Colors.white,
+            size: 18,
+          ),
+        ],
+      ),
+    );
+  }
+}
+
+class StorageIcon extends StatelessWidget {
+  final AssetState storage;
+
+  const StorageIcon({super.key, required this.storage});
+
+  @override
+  Widget build(BuildContext context) {
+    return switch (storage) {
+      AssetState.local => const Positioned(
+          right: 8,
+          bottom: 5,
+          child: Icon(
+            Icons.cloud_off_outlined,
+            color: Color.fromRGBO(255, 255, 255, 0.8),
+            size: 16,
+            shadows: [
+              Shadow(
+                blurRadius: 5.0,
+                color: Color.fromRGBO(0, 0, 0, 0.6),
+                offset: Offset(0.0, 0.0),
+              ),
+            ],
+          ),
+        ),
+      AssetState.remote => const Positioned(
+          right: 8,
+          bottom: 5,
+          child: Icon(
+            Icons.cloud_outlined,
+            color: Color.fromRGBO(255, 255, 255, 0.8),
+            size: 16,
+            shadows: [
+              Shadow(
+                blurRadius: 5.0,
+                color: Color.fromRGBO(0, 0, 0, 0.6),
+                offset: Offset(0.0, 0.0),
+              ),
+            ],
+          ),
+        ),
+      AssetState.merged => const Positioned(
+          right: 8,
+          bottom: 5,
+          child: Icon(
+            Icons.cloud_done_outlined,
+            color: Color.fromRGBO(255, 255, 255, 0.8),
+            size: 16,
+            shadows: [
+              Shadow(
+                blurRadius: 5.0,
+                color: Color.fromRGBO(0, 0, 0, 0.6),
+                offset: Offset(0.0, 0.0),
+              ),
+            ],
+          ),
+        ),
+    };
+  }
+}
+
+class ImageIcon extends StatelessWidget {
```

Comment:
These aren't used anywhere else in the codebase. So they could probably be private widgets instead

```suggestion
class _ImageIcon extends StatelessWidget {
```